zerver/tests/test_link_embed.py

```diff
@@ -309,3 +309,33 @@ def test_invalid_link(self):
         with self.settings(INLINE_URL_EMBED_PREVIEW=True, TEST_SUITE=False, CACHES=TEST_CACHES):
             self.assertIsNone(get_link_embed_data('com.notvalidlink'))
             self.assertIsNone(get_link_embed_data(u'μένει.com.notvalidlink'))
+
+    @override_settings(INLINE_URL_EMBED_PREVIEW=True)
+    def test_message_update_race_condition(self):
+        # type: () -> None
+        url = 'http://test.org/'
+        msg_id = self.send_personal_message(
+            self.example_email('prospero'),
+            self.example_email('cordelia'),
+            content=url,
+        )
+        msg = Message.objects.select_related("sender").get(id=msg_id)
+        # Mock the network request result so the test can be fast without Internet
+        response = MockPythonResponse(self.open_graph_html, 200)
+        mocked_response = mock.Mock(
+            side_effect=lambda k: {url: response}.get(k, MockPythonResponse('', 404)))
+        # Run the queue processor to potentially rerender things
+        with self.settings(TEST_SUITE=False, CACHES=TEST_CACHES):
+            with mock.patch('requests.get', mocked_response):
+                event = {
+                    'message_id': msg_id,
+                    'urls': [url],
+                    'message_realm_id': msg.sender.realm_id,
+                    'message_content': "Not the original content 'http://test.org/'."}
+                FetchLinksEmbedData().consume(event)
```

Comment:
I'm not sure I understand this test, but I guess I'd have thought that what we want is to actually use the message-editing API...

Reply:
Hm, so I thought what we want to test is this case:
1. Push an event with some message to the FetchLinksEmbedData queue.
2. Before event gets handled, update_message gets called. update_message pushes a new event to the queue.
3. First event gets handled, has outdated message.
4. Event gets dropped
5. Second event gets handled.

This test is incomplete in that it only tests until step 4, and does so by inserting an event with the "outdated" content `Not the original content 'http://test.org/'.`. I agree it would be more useful to have a full flow with 
making a call to `client_patch('message')`, but I didn't see a straightforward way to simulate the racing queue_json_publish calls. I guess I could try smt with mocking and adding side_effects?

Reply:
Yeah, I would just mock `queue_json_publish` for the first thing, capturing the event payload, then do `update_message`.  And then finish as you have it here, call `consume` on the captured event now that the edit has gone through.